### PR TITLE
Refactor build scripts into Makefile

### PR DESCRIPTION
**Pull request by mellenburg**
This PR intends to decrease the complexity of the build scripts in this repo. Before this PR, the flow involved multiple circular dependencies with the scripts in `bin/`. Additionally, much of the logic was split and redefined across the multitude of shell scripts. Finally, to run the tests you also had to previously download a repo and compile a jar and it was completely undocumented.

With this new setup, one may completely build and test any kind of spark distribution.
Features:
* If devs wish to use their local environment to compile, thats fine. However, there is now included a Dockerfile (can be made with `make docker-build`) in which devs can execute the build or tests. Note: building the docker dist image will require running outside of the docker container.
* Steps can be easily repeated without tedious setup through the use of targets.
* no more heavy dependency on external tooling. Just pytest, virtualenv, dcos-launch, and a simple login script

Outstanding issues:
* needs dcos-launch installed in the local environment (included in Dockerfile)
* needs better integration with release scripts (still dependent upon jenkins)

### CODE REVIEW

**Comment by elezar** [comment]
@mellenburg would it make sense to rather align this with how other packages (e.g. Confluent) are built and packaged? There are certain features there, e.g. automatically including the `upgradesFrom` and `downgradesTo` versions which may be useful in future.

**Comment by mellenburg** [comment]
@elezar I'm afraid I don't quite understand what you are suggesting... I'm touching none of the code that does the packaging itself. Rather, I am just trying to consolidate the coordination of all these build targets as well as cut out stuff we do not want to maintain. If your suggestion is that this should use "build_frameworks" script, then that seems like something we can do after this change, no?

**Comment by elezar** [comment]
Sorry, I may have misunderstood the intent of the PR. Removing things that we don't want to maintain is definitely good, and I also like using Docker to try and manage external dependencies.

I think @ArtRand has a better handle on how things should be done in this repo, and will be a better reviewer.

File: README.md [hunk 1]
```diff
@@ -9,52 +9,116 @@ described below.
...
+This will setup a docker image that can be used to build the spark
+distribution, the stub univserse, and run the integration test.
+
+This image will not be invoked by any tests explicitly and is not
+required. Rather, it provides a consistent and portable environment for
+`make` to be run in.
 
+
+### Spark Distribution
+A spark distribution is required to create a spark framework package.
```
Comment (elezar) [inline README.md]: Should "spark" be "Apache Spark" (/cc @sascala) throughout?

File: README.md [hunk 2]
```diff
@@ -9,52 +9,116 @@ described below.
...
 
-This is called when you build a Spark package, but sometimes you might
-wish to just create a docker image, such as when you want to test a
-new Spark distribution in client mode, rather than in cluster mode
-through the Dispatcher.
+```bash
+make docker-login
+make docker-dist
+```
+This command will build and push the docker image that will be used by
```
Comment (elezar) [inline README.md]: In text "docker" should be "Docker"

File: bin/jenkins-dist-publish.sh [hunk 3]
```diff
@@ -5,6 +5,37 @@
...
+
+# rename build/dist/spark-*.tgz to build/dist/spark-<TAG>.tgz
+# globals: $SPARK_VERSION
+function rename_dist {
+    SPARK_DIST_DIR="spark-${SPARK_VERSION}-bin-${HADOOP_VERSION}"
+    SPARK_DIST="${SPARK_DIST_DIR}.tgz"
+
+    mkdir -p "${SPARK_DIST_DIR}"
+
+    pushd "${DIST_DIR}"
```
Comment (elezar) [inline bin/jenkins-dist-publish.sh]: Where is `DIST_DIR` defined?

File: bin/jenkins-dist-publish.sh [hunk 4]
```diff
@@ -5,6 +5,37 @@
...
+    tar xvf spark-*.tgz
+    rm spark-*.tgz
```
Comment (elezar) [inline bin/jenkins-dist-publish.sh]: Can one use `tar --transform` here (https://www.gnu.org/software/tar/manual/html_section/tar_52.html) instead of extracting and then recreating the archive?

File: tools/dcos_login.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import os
+
+import dcos.cluster
+import requests
+
+from dcos_test_utils import logger
+
+__CLI_LOGIN_OPEN_TOKEN = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik9UQkVOakZFTWtWQ09VRTRPRVpGTlRNMFJrWXlRa015Tnprd1JrSkVRemRCTWpBM1FqYzVOZyJ9.eyJlbWFpbCI6ImFsYmVydEBiZWtzdGlsLm5ldCIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJpc3MiOiJodHRwczovL2Rjb3MuYXV0aDAuY29tLyIsInN1YiI6Imdvb2dsZS1vYXV0aDJ8MTA5OTY0NDk5MDExMTA4OTA1MDUwIiwiYXVkIjoiM3lGNVRPU3pkbEk0NVExeHNweHplb0dCZTlmTnhtOW0iLCJleHAiOjIwOTA4ODQ5NzQsImlhdCI6MTQ2MDE2NDk3NH0.OxcoJJp06L1z2_41_p65FriEGkPzwFB_0pA9ULCvwvzJ8pJXw9hLbmsx-23aY2f-ydwJ7LSibL9i5NbQSR2riJWTcW4N7tLLCCMeFXKEK4hErN2hyxz71Fl765EjQSO5KD1A-HsOPr3ZZPoGTBjE0-EFtmXkSlHb1T2zd0Z8T5Z2-q96WkFoT6PiEdbrDA-e47LKtRmqsddnPZnp0xmMQdTr2MjpVgvqG7TlRvxDcYc-62rkwQXDNSWsW61FcKfQ-TRIZSf2GS9F9esDF4b5tRtrXcBNaorYa9ql0XAWH5W_ct4ylRNl3vwkYKWa4cmPvOqT5Wlj9Tf0af4lNO40PQ'  # noqa
+__CLI_LOGIN_EE_USERNAME = 'bootstrapuser'
+__CLI_LOGIN_EE_PASSWORD = 'deleteme'
```
Comment (elezar) [inline tools/dcos_login.py]: Should this file contain credentials?

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    python3-dev \
+    python3-pip \
+    python3-venv \
+    python-software-properties \
+    software-properties-common \
+    wget \
+    zip && \
+    rm -rf /var/lib/apt/lists/*
+# install go 1.7
+RUN add-apt-repository -y ppa:longsleep/golang-backports && apt-get update && apt-get install -y golang-go
```
[the remark this answers is not in the document]
Comment (mellenburg) [reply Dockerfile]: dcos-commons would actually fail not too long ago if go 1.7 wasnt included. This was just copy paste from there so we can just change it to 1.8 if desired

File: Makefile [hunk 7]
```diff
@@ -1,13 +1,220 @@
...
+	popd
+
+HADOOP_VERSION := $(shell jq ".default_spark_dist.hadoop_version" "$(ROOT_DIR)/manifest.json")
+
+SPARK_DIR := $(ROOT_DIR)/spark
+$(SPARK_DIR):
+	git clone https://github.com/mesosphere/spark $(SPARK_DIR)
+
+# Builds a quick dev version of spark from the mesosphere fork
+dev-dist: $(SPARK_DIR)
```
[the remark this answers is not in the document]
Comment (mellenburg) [reply Makefile]: Yeah, that is exactly the intent. However, when variables are set like this, you need to do an *override*, not just set it in the env:
`make dev-dist -e SPARK_DIR=xyz`

File: Makefile [hunk 8]
```diff
@@ -1,13 +1,220 @@
...
+	UNIVERSE_URL_PATH=$(UNIVERSE_URL_PATH) \
+	TEMPLATE_CLI_VERSION=$(CLI_VERSION) \
+	TEMPLATE_DOCKER_IMAGE=`cat docker-dist` \
+		$(TOOLS_DIR)/publish_aws.py \
+		spark \
+        $(ROOT_DIR)/universe/ \
+        $(CLI_DIST_DIR)/dcos-spark-darwin \
+        $(CLI_DIST_DIR)/dcos-spark-linux \
+        $(CLI_DIST_DIR)/dcos-spark.exe \
+        $(CLI_DIST_DIR)/*.whl; \
```
[the remark this answers is not in the document]
Comment (mellenburg) [reply Makefile]: Nope, will cleanup

File: Makefile [hunk 9]
```diff
@@ -1,13 +1,220 @@
...
+	cp -r conf/* $(BUILD_DIR)/docker/dist/conf
+	cp -r docker/* $(BUILD_DIR)/docker
+	pushd $(BUILD_DIR)/docker
+	docker build -t $(DOCKER_DIST_IMAGE) .
+	popd
+	docker push $(DOCKER_DIST_IMAGE)
+	echo "$(DOCKER_DIST_IMAGE)" > $@
+
+CLI_VERSION := $(shell jq -r ".cli_version" "$(ROOT_DIR)/manifest.json")
+$(CLI_DIST_DIR):
```
[the remark this answers is not in the document]
Comment (mellenburg) [reply Makefile]: Good catch. I was just doing this to be lazy. Will add phony target

File: README.md [hunk 10]
```diff
@@ -9,52 +9,116 @@ described below.
...
+```bash
+make test-env
+```
+This command will setup a virtual environment that will contain the CLI and
+other requirements for running the pytest integration suite
+
+```bash
+make cluster_info.json
+```
+This command will make a cluster that can be used for integration testing.
```
[the remark this answers is not in the document]
Comment (mellenburg) [reply README.md]: Oops, its a typo that should be removed. Good catch!

File: tools/dcos_login.py (hunk 5 above)
Comment (mellenburg) [reply tools/dcos_login.py]: This is the same as dcos-commons. These are default credentials that have been basically public for a long time

File: bin/jenkins-dist-test.sh [hunk 11]
```diff
@@ -1,17 +0,0 @@
-#!/bin/bash
```
[the remark this answers is not in the document]
Comment (mellenburg) [reply bin/jenkins-dist-test.sh]: @susanxhuynh This script is pretty trivial, can we just change the job? Part of the motivation for this work is to reduce a bunch of extra entrypoints that do nothing special. The flow of this make file makes it extremely easy to do exactly this.

Also, we do want to move all the things to TeamCity so we don't have projects split across two CIs and can better track costs, so lets just move this?

[the remark this answers is not in the document]
Comment (susanxhuynh) [reply bin/jenkins-dist-test.sh]: @mellenburg Yes, moving this job to TeamCity makes sense. Let's plan on moving it.

File: Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (susanxhuynh) [reply Dockerfile]: Eh, I just wanted to make sure it aligned with the data services frameworks (Cassandra, Kafka, etc).

File: tools/dcos_login.py (hunk 5 above)
Comment (elezar) [reply tools/dcos_login.py]: Yes, I've since noticed it elsewhere. If this is intentional, then my comment can be ignored.

File: Makefile [hunk 12]
```diff
@@ -1,13 +1,220 @@
...
+docker-dist: $(DIST_DIR)
+	tar xvf $(DIST_DIR)/spark-*.tgz -C $(DIST_DIR)
+	rm -rf $(BUILD_DIR)/docker
+	mkdir -p $(BUILD_DIR)/docker/dist
+	cp -r $(DIST_DIR)/spark-*/. $(BUILD_DIR)/docker/dist
+	cp -r conf/* $(BUILD_DIR)/docker/dist/conf
+	cp -r docker/* $(BUILD_DIR)/docker
+	pushd $(BUILD_DIR)/docker
+	docker build -t $(DOCKER_DIST_IMAGE) .
+	popd
```
[the remark this answers is not in the document]
Comment (susanxhuynh) [reply Makefile]: ^^ Might be a problem on OS X but not Linux.

File: Makefile [hunk 13]
```diff
@@ -1,13 +1,220 @@
...
+	echo $(DOCKER_BUILD_IMAGE) > $@
 
-.PHONY: dist docker test universe
+# Pulls the spark distribution listed in the manifest as default
+SPARK_DIST_URI := $(shell jq ".default_spark_dist.uri" "$(ROOT_DIR)/manifest.json")
+manifest-dist:
+	mkdir -p $(DIST_DIR)
+	pushd $(DIST_DIR)
+	wget $(SPARK_DIST_URI)
+	popd
```
[the remark this answers is not in the document]
Comment (mellenburg) [reply Makefile]: I have ONESHELL set in this script so it actually works just fine. Is that a blocker to require make 4?

**Comment by mellenburg** [comment]
@elezar Sorry that this comment comes so long after your feedback

> It seems as if the Makefile here is dependent not only on the structure of this repository, but on the structure of the spark repository itself (where we use mesosphere/spark, for example). 

The scripts used here before this change make even stronger assumptions about how repos should be structured. Nowhere in here does it say that these scripts should only be invoked from one level outside the repo (the repo MUST be named spark-build too) and that there needs to be spark repo along side it which may or may not be the mesosphere fork. What I do here is specifically parameterize these magic directories such that one may (1) specify preexisting directories, whether it be the fork or not (2) if those directories do not exist on the system running the makefile, a sensible default will be used automatically.


> This seems to add a lot of make commands (docker-dist for example) which actually build Docker images from the Spark distribution. 

The number of make commands is not more than the number of scripts present in the bin/ folder. `docker-dist` is already present, just as a one-off script....which is literally what a make recipe is. These rely on a massive shared bash file, but some of them inconsistently override values from there.  In other cases, the make file references a bin/* script which in turn... CALLS THE MAKEFILE AGAIN. Trying to figure out where inconsistencies or bugs are in such a setup is quite agonizing. To be fair, there is one "new" target: `mesos-spark-integration-tests` but this is currently handled as a jenkins artifact that is magically inserted into the build. This MUST be documented and executed by checked in code.
TL;DR: there are no extra commands really being added here. They are just being transferred from bin/* scripts or places that were inconsistently implemented

> Would it not make more sense to split these concerns, and have a pre-built docker image as the input for the jobs here. The same argument follows for building Spark from source.

I did not start this repo, but it seems to me the entire intent of this repo is building the spark framework from a spark binary, which in turn requires a docker container to be injected into the final framework. As such, the steps used to create the container and binary must be strictly controlled here. Note, you CAN use a prebuilt docker image. Just override it. If you don't happen to know what magic prebuilt docker image should be used. This Makefile is nice enough to make it for you. 

> Then, as I have asked before, this repo contains (some of) the tooling from the dcos-commons repository, but the utilities that we have there (like the universe builder) do not seem to be used

This is out of my scope. I am just copy/pasting the methods currently used. As far as I know, there isnt some particular way to build these frameworks that live outside of the monorepo

File: Makefile (hunk 13 above)
[the remark this answers is not in the document]
Comment (susanxhuynh) [reply Makefile]: No, I don't think it's a blocker. The `make` version on my Mac is still on 3.81. I guess I could upgrade it.

**Comment by susanxhuynh** [comment]
@mellenburg I think that one test failure (test_s3) will pass if you rebase.

**Comment by mellenburg** [comment]
d'oh. Thanks susan!
